arch/risc-v/src/common/riscv_addrenv_util.c

```diff
@@ -0,0 +1,57 @@
+/****************************************************************************
+ * arch/risc-v/src/common/riscv_addrenv_util.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <arch/board/board_memorymap.h>
+
+#include "pgalloc.h"
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: riscv_pgpool_to_vaddr
+ *
+ * Description:
+ *   Convert physical page pool address to virtual page pool address
+ *
+ * Input Parameters:
+ *   paddr - The physical address to convert
+ *
+ * Returned Value:
+ *   Page pool virtual address, or 0 if paddr is not within page pool
+ *
+ ****************************************************************************/
+
+uintptr_t riscv_pgpool_to_vaddr(uintptr_t paddr)
```

Comment:
why not follow arm_pgvaddr implementation? I mean:

1. Name it riscv_pgvaddr
2. Use CONFIG_ARCH_PGPOOL_PBASE and CONFIG_ARCH_PGPOOL_VBASE
3. Implement as macro

Reply:
I can change the name but I don't want to use the CONFIG_ARCH_PGPOOL_PBASE/VBASE values. I can if you insist, but I already give the values as linker defined symbols, so doing the same thing twice did not make sense to me.

Reply:
ARM actually seems to have a double implementation for the same thing basically:
static inline uintptr_t arm_pgvaddr(uintptr_t paddr);
uintptr_t arm_virtpgaddr(uintptr_t paddr);

The first is maybe used to get the kernel virtual address for a page pool page.
The second is maybe used to get the user mapped address for a page pool page.

I'll do it like the first prototype, as the purpose of the function is to get the kernel mapping

Reply:
But ARCH_PGPOOL_PBASE is defined in arch/Kconfig, and then shared by all arch(both arm and riscv), which mean the user could change the setting through ARCH_PGPOOL_PBASE, but you ignore it totally. It isn't good.

As I suggest before, if you think other arch implementation isn't good enough and you have good idea to improve it. Please improve other arch first, otherwise follow the old implementation.

Reply:
For example, the implementation of fputest isn't good, @anchao(#6080) provide a patch not only fix arm, but also riscv. We need consider the project as whole and all components equally, an improve them together as much as we can. Another example, you find address environment issue and fixed in #5913, even we never use the kernel mode in the production, but we still take additional time to bring the change back to other arch: #6089, #6089 and #6099.

Reply:
I think there are issues with both approaches. One can select a Kconfig value that is wrong, no matter what I do. One can select a virtual address mapping for PGPOOL_VBASE that is overlapping with another virtual address, or not addressable at all. One can select a PGPOOL_PBASE value that is e.g. in kernel code memory, or not in any addressable memory at all.

I think fundamentally the location of the page pool can be anywhere. I personally prefer using linker defined symbols to indicate memory locations, as then:
a) The linker will not accidentally put anything else there
b) The linker provided symbols for the memory region will always be correct

But this is a philosophical dispute, linker is used to place symbols, objects do relocations etc. It is not its responsibility to define memory areas. However it has to know something about the memory layout, so for now it DOES define memory areas.

One option would be to add assert tests to ensure that the ARCH_PGPOOL_PBASE value at least matches the PGPOOL_START value. The ARCH_PGPOOL_VBASE value can still be wrong (overlap with other areas) but there is no test to ensure that.

Reply:
> For example, the implementation of fputest isn't good, @anchao(#6080) provide a patch not only fix arm, but also riscv. We need consider the project as whole and all components equally, and improve them together as much as we can. Another example, you find address environment issue which is fixed by #5913, even we never use the kernel mode in the production, but we still take additional time to bring the change back to other arch: #6089, #6089 and #6099.

Yes I think it's great you are patching both platforms, and especially the unifying work under risc-v that has removed a massive amount of copy & paste is fantastic.

But please keep in mind, I'm doing this on my own and not capable of doing as much as you guys.

Reply:
I did the modification as you suggested